src/core/server/http/integration_tests/request.test.ts

```diff
@@ -329,4 +330,63 @@ describe('KibanaRequest', () => {
       expect(resp1.body.requestUuid).toBe('xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx');
     });
   });
+
+  describe('issue-73849', () => {
+    it('headers timeout is handled correctly', async () => {
+      const { server: innerServer, createRouter } = await server.setup(setupDeps);
+      const router = createRouter('/');
+
+      router.get({ path: '/', validate: false }, async (context, request, res) => {
+        return res.ok({ body: 'ok' });
+      });
+
+      await server.start();
+
+      const agent = new Http.Agent({
+        keepAlive: true,
+      });
+
+      const oneSec = 1_000;
+
+      function performRequest() {
+        return new Promise((resolve, reject) => {
+          const req = Http.request(
+            {
+              protocol: 'http:',
+              host: 'localhost',
+              port: 5601,
+              path: '/',
+              method: 'GET',
+              agent,
+            },
+            function (res) {
+              let data = '';
+              res.on('data', (chunk) => {
+                data += chunk;
+              });
+              res.on('end', () => resolve(data));
+            }
+          );
+
+          req.on('socket', (socket) => {
+            socket.write('GET / HTTP/1.1\r\n');
+            setTimeout(() => {
+              socket.write('Host: localhost\r\n');
+            }, oneSec);
+            setTimeout(() => {
+              // headersTimeout doesn't seem to fire if request headers are sent in one packet.
+              socket.write('\r\n');
+              req.end();
+            }, 2 * oneSec);
+          });
+
+          req.on('error', reject);
+        });
+      }
+
+      await performRequest();
+      await delay(innerServer.listener.headersTimeout + oneSec);
```

Comment:
It makes the test last long, but I'm uncomfortable to test such things with mocked timers. Furthermore, we are going to remove this code and the test in the next minor version release.